File: Data Structures & Algorithms/combination-target-sum/submission-1.kt
class Solution {
    fun combinationSum(nums: IntArray, target: Int): List<List<Int>> {
        val result = mutableListOf<List<Int>>()

        fun dfs(index: Int, total: Int, current: MutableList<Int>) {
            if(total == target) {
                result.add(ArrayList(current))
                return
            }
            if(index >= nums.size || total > target) {
                return
            }

            current.add(nums[index])
            dfs(index, total + nums[index], current)
            current.removeLast()
            dfs(index + 1, total, current)
        }

        dfs(0, 0, mutableListOf())
        return result
    }
}

File: Data Structures & Algorithms/two-integer-sum/submission-1.cpp
class Solution {
   public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> map;
        for (int i = 0; i < nums.size(); i++) {
            int compelment = target - nums[i];
            if (map.count(compelment) >= 1) {
                return { map[compelment], i };
            }
            map[nums[i]] = i;
        }

        return {};
    }
};
